src/mongo_odm/query_builder.hpp

```diff
@@ -25,93 +26,453 @@
 #include <bsoncxx/view_or_value.hpp>
 
 #include <mongo_odm/nvp.hpp>
+#include <mongo_odm/util.hpp>
 
 namespace mongo_odm {
 MONGO_ODM_INLINE_NAMESPACE_BEGIN
 
-// Forward declarations
-template <typename Base, typename T>
-class NotExpr;
+// Forward declarations for Expression type trait structs
+template <typename>
+struct is_sort_expression;
+template <typename>
+struct is_query_expression;
+template <typename>
+struct is_update_expression;
+
+/**
+ * Type trait that contains true if a type T can be appended to a BSON builder using
+ * builder.append(T val);
+ */
+template <typename T>
+constexpr auto is_bson_appendable_impl(int)
+    -> decltype(std::declval<bsoncxx::builder::core>().append(std::declval<T>()),
+                std::true_type{}) {
+    return {};
+}
+
+template <typename T>
+constexpr std::false_type is_bson_appendable_impl(...) {
+    return {};
+}
+
+template <typename T>
+using is_bson_appendable = decltype(is_bson_appendable_impl<T>(0));
```

Comment:
I meant that you could leave the definition of is_bson_appendable the same and remove the impl definitions (just leaving the declarations) and it would all still work.

You get truthy or falsey-ness just from the return type, which is determined just from overload selection (not actually running anything).